app/src/main/java/com/example/androidtraining/MainActivity.kt

```diff
@@ -37,10 +84,141 @@ fun Greeting(name: String, modifier: Modifier = Modifier) {
+@OptIn(ExperimentalMaterial3Api::class)
 @Composable
-fun GreetingPreview() {
-    AndroidTrainingTheme {
-        Greeting("Android")
+fun HomeScreen(navigateToProductScreen: () -> Unit = {}) {
+    var email = remember { mutableStateOf("") }
+    var password = remember { mutableStateOf("") }
+    var passwordVisible = remember { mutableStateOf(false) }
+    var errorVisible = remember { mutableStateOf(false) }
+    Column(
+        modifier = Modifier
+            .fillMaxSize()
+            .padding(horizontal = 20.dp, vertical = 20.dp)
+    ) {
+        Row(
+            modifier = Modifier
+                .fillMaxWidth()
+                .padding(vertical = 20.dp),
+            horizontalArrangement = Arrangement.Center
+        ) {
+            Image(
+                painter = painterResource(id = R.drawable.logo),
+                contentDescription = "logo",
+                modifier = Modifier
+                    .size(200.dp)
+            )
+        }
+        Text(text = "Log in", fontWeight = FontWeight.Bold, fontSize = 30.sp)
+        OutlinedTextField(
+            value = email.value,
+            onValueChange = { newValue ->
+                email.value = newValue
+            },
+            isError = !isValidEmail(email.value),
+            label = { Text("Email") },
+            placeholder = { Text("Input email") },
+            modifier = Modifier
+                .fillMaxWidth()
+        )
+        OutlinedTextField(value = password.value,
+            onValueChange = { password.value = it },
+            label = { Text("Password") },
+            singleLine = true,
+            placeholder = { Text("Password") },
+            modifier = Modifier
+                .fillMaxWidth(),
+            visualTransformation = if (passwordVisible.value) VisualTransformation.None else PasswordVisualTransformation(),
+            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password),
+            trailingIcon = {
+                val image = if (passwordVisible.value)
+                    painterResource(R.drawable.visibility)
+                else painterResource(R.drawable.visibility_off)
+
+                val description = if (passwordVisible.value) "Hide password" else "Show password"
+
+                IconButton(onClick = { passwordVisible.value = !passwordVisible.value }) {
+                    Image(painter = image, description)
+                }
+            })
+        Button(modifier = Modifier
+            .fillMaxWidth()
+            .padding(vertical = 5.dp, horizontal = 10.dp),
+            onClick = {
+                if (checkForValidCredentials(
+                        email.value,
+                        password.value
+                    )
+                ) navigateToProductScreen()
+                else errorVisible.value = true
+            }
+        ) {
+            Text(text = "Log in")
+        }
+        Text(text = if (errorVisible.value) "Invalid credentials" else "")
```

Comment:
Prefer to wrap the whole Composable instead of the text that needs to be shown -- this way less items will be rendered on the screen and you won't have mysterious white spaces if you had something under this `Text`.

Reply:
fixed